File: src/xianyu_cli/core/crypto.py
"""MessagePack decoder for XianYu WebSocket messages.

Decodes Base64-encoded MessagePack binary data from syncPushPackage.
This is a minimal pure-Python implementation — no external msgpack dependency.
"""

from __future__ import annotations

import base64
import struct
from typing import Any


class MessagePackDecoder:
    """Pure Python MessagePack decoder for XianYu WebSocket push messages."""

    def __init__(self, data: bytes):
        self.data = data
        self.pos = 0

    def read_byte(self) -> int:
        if self.pos >= len(self.data):
            raise ValueError("Unexpected end of data")
        byte = self.data[self.pos]
        self.pos += 1
        return byte

    def read_bytes(self, n: int) -> bytes:
        if self.pos + n > len(self.data):
            raise ValueError("Unexpected end of data")
        result = self.data[self.pos : self.pos + n]
        self.pos += n
        return result

    def read_uint16(self) -> int:
        return struct.unpack(">H", self.read_bytes(2))[0]

    def read_uint32(self) -> int:
        return struct.unpack(">I", self.read_bytes(4))[0]

    def read_int8(self) -> int:
        return struct.unpack(">b", self.read_bytes(1))[0]

    def read_int16(self) -> int:
        return struct.unpack(">h", self.read_bytes(2))[0]

    def read_int32(self) -> int:
        return struct.unpack(">i", self.read_bytes(4))[0]

    def read_float32(self) -> float:
        return struct.unpack(">f", self.read_bytes(4))[0]

    def read_float64(self) -> float:
        return struct.unpack(">d", self.read_bytes(8))[0]

    def read_string(self, length: int) -> str:
        return self.read_bytes(length).decode("utf-8", errors="replace")

    def decode_value(self) -> Any:
        fmt = self.read_byte()

        # Positive fixint (0x00 - 0x7f)
        if fmt <= 0x7F:
            return fmt
        # Fixmap (0x80 - 0x8f)
        if 0x80 <= fmt <= 0x8F:
            return self.decode_map(fmt & 0x0F)
        # Fixarray (0x90 - 0x9f)
        if 0x90 <= fmt <= 0x9F:
            return self.decode_array(fmt & 0x0F)
        # Fixstr (0xa0 - 0xbf)
        if 0xA0 <= fmt <= 0xBF:
            return self.read_string(fmt & 0x1F)
        # Nil
        if fmt == 0xC0:
            return None
        # False
        if fmt == 0xC2:
            return False
        # True
        if fmt == 0xC3:
            return True
        # bin 8
        if fmt == 0xC4:
            length = self.read_byte()
            return self.read_bytes(length)
        # bin 16
        if fmt == 0xC5:
            length = self.read_uint16()
            return self.read_bytes(length)
        # bin 32
        if fmt == 0xC6:
            length = self.read_uint32()
            return self.read_bytes(length)
        # float 32
        if fmt == 0xCA:
            return self.read_float32()
        # float 64
        if fmt == 0xCB:
            return self.read_float64()
        # uint 8
        if fmt == 0xCC:
            return self.read_byte()
        # uint 16
        if fmt == 0xCD:
            return self.read_uint16()
        # uint 32
        if fmt == 0xCE:
            return self.read_uint32()
        # int 8
        if fmt == 0xD0:
            return self.read_int8()
        # int 16
        if fmt == 0xD1:
            return self.read_int16()
        # int 32
        if fmt == 0xD2:
            return self.read_int32()
        # str 8
        if fmt == 0xD9:
            length = self.read_byte()
            return self.read_string(length)
        # str 16
        if fmt == 0xDA:
            length = self.read_uint16()
            return self.read_string(length)
        # str 32
        if fmt == 0xDB:
            length = self.read_uint32()
            return self.read_string(length)
        # array 16
        if fmt == 0xDC:
            return self.decode_array(self.read_uint16())
        # array 32
        if fmt == 0xDD:
            return self.decode_array(self.read_uint32())
        # map 16
        if fmt == 0xDE:
            return self.decode_map(self.read_uint16())
        # map 32
        if fmt == 0xDF:
            return self.decode_map(self.read_uint32())
        # Negative fixint (0xe0 - 0xff)
        if fmt >= 0xE0:
            return fmt - 256

        return None

    def decode_map(self, size: int) -> dict[str, Any]:
        result: dict[str, Any] = {}
        for _ in range(size):
            key = self.decode_value()
            value = self.decode_value()
            result[str(key) if not isinstance(key, str) else key] = value
        return result

    def decode_array(self, size: int) -> list[Any]:
        return [self.decode_value() for _ in range(size)]


def decrypt_message(data: str) -> Any:
    """Decrypt a WebSocket push message.

    Flow: Base64 decode → MessagePack decode → parsed object
    """
    try:
        raw = base64.b64decode(data)
        decoder = MessagePackDecoder(raw)
        return decoder.decode_value()
    except Exception:
        # Fallback: try direct JSON
        import json
        import logging

        logging.getLogger(__name__).debug("MessagePack decode failed, trying JSON")
        try:
            return json.loads(data)
        except (json.JSONDecodeError, TypeError):
            logging.getLogger(__name__).debug("JSON fallback also failed, returning raw")
            return data

File: golang/core/crypto.go
package core

import (
	"encoding/base64"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"math"
)

// MessagePackDecoder decodes MessagePack binary data.
type MessagePackDecoder struct {
	data []byte
	pos  int
}

// NewMessagePackDecoder creates a decoder from raw bytes.
func NewMessagePackDecoder(data []byte) *MessagePackDecoder {
	return &MessagePackDecoder{data: data}
}

func (d *MessagePackDecoder) readByte() (byte, error) {
	if d.pos >= len(d.data) {
		return 0, fmt.Errorf("unexpected end of data")
	}
	b := d.data[d.pos]
	d.pos++
	return b, nil
}

func (d *MessagePackDecoder) readBytes(n int) ([]byte, error) {
	if d.pos+n > len(d.data) {
		return nil, fmt.Errorf("unexpected end of data")
	}
	result := d.data[d.pos : d.pos+n]
	d.pos += n
	return result, nil
}

func (d *MessagePackDecoder) readUint16() (uint16, error) {
	b, err := d.readBytes(2)
	if err != nil {
		return 0, err
	}
	return binary.BigEndian.Uint16(b), nil
}

func (d *MessagePackDecoder) readUint32() (uint32, error) {
	b, err := d.readBytes(4)
	if err != nil {
		return 0, err
	}
	return binary.BigEndian.Uint32(b), nil
}

func (d *MessagePackDecoder) readString(length int) (string, error) {
	b, err := d.readBytes(length)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

// DecodeValue decodes the next MessagePack value.
func (d *MessagePackDecoder) DecodeValue() (interface{}, error) {
	b, err := d.readByte()
	if err != nil {
		return nil, err
	}

	// Positive fixint (0x00 - 0x7f)
	if b <= 0x7F {
		return int(b), nil
	}
	// Fixmap (0x80 - 0x8f)
	if b >= 0x80 && b <= 0x8F {
		return d.decodeMap(int(b & 0x0F))
	}
	// Fixarray (0x90 - 0x9f)
	if b >= 0x90 && b <= 0x9F {
		return d.decodeArray(int(b & 0x0F))
	}
	// Fixstr (0xa0 - 0xbf)
	if b >= 0xA0 && b <= 0xBF {
		return d.readString(int(b & 0x1F))
	}
	// Nil
	if b == 0xC0 {
		return nil, nil
	}
	// False
	if b == 0xC2 {
		return false, nil
	}
	// True
	if b == 0xC3 {
		return true, nil
	}
	// bin 8
	if b == 0xC4 {
		length, err := d.readByte()
		if err != nil {
			return nil, err
		}
		return d.readBytes(int(length))
	}
	// bin 16
	if b == 0xC5 {
		length, err := d.readUint16()
		if err != nil {
			return nil, err
		}
		return d.readBytes(int(length))
	}
	// bin 32
	if b == 0xC6 {
		length, err := d.readUint32()
		if err != nil {
			return nil, err
		}
		return d.readBytes(int(length))
	}
	// float 32
	if b == 0xCA {
		bs, err := d.readBytes(4)
		if err != nil {
			return nil, err
		}
		bits := binary.BigEndian.Uint32(bs)
		return math.Float32frombits(bits), nil
	}
	// float 64
	if b == 0xCB {
		bs, err := d.readBytes(8)
		if err != nil {
			return nil, err
		}
		bits := binary.BigEndian.Uint64(bs)
		return math.Float64frombits(bits), nil
	}
	// uint 8
	if b == 0xCC {
		v, err := d.readByte()
		return int(v), err
	}
	// uint 16
	if b == 0xCD {
		v, err := d.readUint16()
		return int(v), err
	}
	// uint 32
	if b == 0xCE {
		v, err := d.readUint32()
		return int(v), err
	}
	// uint 64
	if b == 0xCF {
		bs, err := d.readBytes(8)
		if err != nil {
			return nil, err
		}
		return int64(binary.BigEndian.Uint64(bs)), nil
	}
	// int 8
	if b == 0xD0 {
		v, err := d.readByte()
		return int(int8(v)), err
	}
	// int 16
	if b == 0xD1 {
		v, err := d.readUint16()
		return int(int16(v)), err
	}
	// int 32
	if b == 0xD2 {
		v, err := d.readUint32()
		return int(int32(v)), err
	}
	// int 64
	if b == 0xD3 {
		bs, err := d.readBytes(8)
		if err != nil {
			return nil, err
		}
		return int64(binary.BigEndian.Uint64(bs)), nil
	}
	// fixext 1
	if b == 0xD4 {
		_, err := d.readBytes(2) // type + 1 byte data
		return nil, err
	}
	// fixext 2
	if b == 0xD5 {
		_, err := d.readBytes(3)
		return nil, err
	}
	// fixext 4
	if b == 0xD6 {
		_, err := d.readBytes(5)
		return nil, err
	}
	// fixext 8
	if b == 0xD7 {
		_, err := d.readBytes(9)
		return nil, err
	}
	// fixext 16
	if b == 0xD8 {
		_, err := d.readBytes(17)
		return nil, err
	}
	// str 8
	if b == 0xD9 {
		length, err := d.readByte()
		if err != nil {
			return nil, err
		}
		return d.readString(int(length))
	}
	// str 16
	if b == 0xDA {
		length, err := d.readUint16()
		if err != nil {
			return nil, err
		}
		return d.readString(int(length))
	}
	// str 32
	if b == 0xDB {
		length, err := d.readUint32()
		if err != nil {
			return nil, err
		}
		return d.readString(int(length))
	}
	// array 16
	if b == 0xDC {
		n, err := d.readUint16()
		if err != nil {
			return nil, err
		}
		return d.decodeArray(int(n))
	}
	// array 32
	if b == 0xDD {
		n, err := d.readUint32()
		if err != nil {
			return nil, err
		}
		return d.decodeArray(int(n))
	}
	// map 16
	if b == 0xDE {
		n, err := d.readUint16()
		if err != nil {
			return nil, err
		}
		return d.decodeMap(int(n))
	}
	// map 32
	if b == 0xDF {
		n, err := d.readUint32()
		if err != nil {
			return nil, err
		}
		return d.decodeMap(int(n))
	}
	// Negative fixint (0xe0 - 0xff)
	if b >= 0xE0 {
		return int(b) - 256, nil
	}

	return nil, nil
}

func (d *MessagePackDecoder) decodeMap(size int) (map[string]interface{}, error) {
	result := make(map[string]interface{})
	for i := 0; i < size; i++ {
		key, err := d.DecodeValue()
		if err != nil {
			return nil, err
		}
		value, err := d.DecodeValue()
		if err != nil {
			return nil, err
		}
		result[fmt.Sprint(key)] = value
	}
	return result, nil
}

func (d *MessagePackDecoder) decodeArray(size int) ([]interface{}, error) {
	result := make([]interface{}, size)
	for i := 0; i < size; i++ {
		v, err := d.DecodeValue()
		if err != nil {
			return nil, err
		}
		result[i] = v
	}
	return result, nil
}

// DecryptMessage decodes a WebSocket push message.
// Flow: Base64 decode → MessagePack decode → parsed object
func DecryptMessage(data string) interface{} {
	raw, err := base64.StdEncoding.DecodeString(data)
	if err != nil {
		// Try direct JSON
		var result interface{}
		if json.Unmarshal([]byte(data), &result) == nil {
			return result
		}
		return data
	}

	decoder := NewMessagePackDecoder(raw)
	result, err := decoder.DecodeValue()
	if err != nil {
		// Try direct JSON
		var jsonResult interface{}
		if json.Unmarshal([]byte(data), &jsonResult) == nil {
			return jsonResult
		}
		return data
	}
	return result
}
